# Books versus eBooks : The customer's choice

You need to have the metadata.json and the two 5-core files reviews_Books_5.json and reviews_Kindle_Store_5.json at the same level as the notebook.

For the questions asked for Milestone 2, please refer to the README file, where you will see a part dedicated to milestone 2. In this Notebook, we don't answer specifically to these questions, but we speak at each step what we technically made and failed.

## Data retrieval

In [11]:
import pandas as pd
import re
import numpy as np
import requests
import time
from bs4 import BeautifulSoup
from ast import literal_eval

#### Warning :
There are several booleans in this cell. They are useful to indicate if we need to perform some key data filtering and transformation in the whole notebook. The first one is set to True as it creates a file that is bigger than 300Mo, so that we can't put it in github. You will have to compute it if you run the notebook (it takes some time). The other files are on github, so all other booleans are set to False.

In [12]:
metadata_path = 'metadata.json'
books_metadata_path = 'books_metadata_with_bracket.csv'
ebooks_metadata_title_path = 'ebooks_metadata_title.csv'
ebooks_asin = 'ebooks_asin.csv'
books_5core_path = 'reviews_Books_5.json'
kindle_5core_path = 'reviews_Kindle_Store_5.json'
amazon_ebooks = 'ebooks_title_from_amazon_complete.csv'
amazon_aws_ebooks = 'ebooks_metadata_amazon.csv'
amazon_aws_ebooks_undefined = 'ebooks_metadata_amazon_undefined.csv'
amazon_aws_ebooks_log = 'ebooks_metadata_amazon_log.csv'
asindb_ebooks = 'ebooks_title_from_asindb.csv'

amazon_aws_books_yves = 'books_metadata_amazon_yves.csv'
amazon_aws_books_pa = 'books_metadata_amazon_pa.csv'

matched_books_path = 'matched_books.csv'
matched_ebooks_path = 'matched_ebooks.csv'
weighted_scores_books_path = 'weighted_scores_books.csv'
weighted_scores_ebooks_path = 'weighted_scores_ebooks.csv'


WRITE_BOOKS_METADATA = False
WRITE_EBOOKS_METADATA_TITLE = False
WRITE_EBOOK_ASIN = False
AMAZON_GET_TITLE = False
ASINDB_GET_TITLE = False
WRITE_FIND_MATCHED = False
WRITE_WEIGHTED_SCORE = True

As for our project, we need to obtain ebook data and book data, we chose the Amazon dataset. On this <a href='http://jmcauley.ucsd.edu/data/amazon/'>link</a>, we have downloaded the Books and Kindle Store 5-core files. However, those files contain reviews, so that we have no information about the article title or price.
For that reason, we had to use the metadata file, acting as an intermediate table (relationship).

We obtained the metadata.json file from the cluster, as it was not available in the website.
We accessed the cluster using ssh@iccluster060.iccluster.epfl.ch, then using 
```shell
hadoop fs -get /datasets/productGraph/metadata.json /buffer
```
to move the dataset to a folder we could connect to with SCP to download it on our computer.


It's a json file, that is not readable using the pandas read_json method. We had to use the Code part from <a href='http://jmcauley.ucsd.edu/data/amazon/'>here</a> to read it. We can see a way to read the file (a limited portion of it) below :

In [12]:
def read_json(path, limit = 2): 
    g = open(path, 'r') 
    df = {}
    for i, l in enumerate(g): 
        if i < limit:
            df[i] = eval(l)
        else:
            break
    return pd.DataFrame.from_dict(df, orient='index')

def read_csv(path, limit = 2): 
    return pd.read_csv(path, nrows=limit)
            
read_json(metadata_path)

asin                salesRank  \
0  0001048791       {'Books': 6334800}   
1  0000143561  {'Movies & TV': 376041}   

                                               imUrl               categories  \
0  http://ecx.images-amazon.com/images/I/51MKP0T4...                [[Books]]   
1  http://g-ecx.images-amazon.com/images/G/01/x-s...  [[Movies & TV, Movies]]   

                                               title  \
0  The Crucible: Performed by Stuart Pankin, Jero...   
1  Everyday Italian (with Giada de Laurentiis), V...   

                                         description  price  \
0                                                NaN    NaN   
1  3Pack DVD set - Italian Classics, Parties and ...  12.99   

                                             related  
0                                                NaN  
1  {'also_viewed': ['B0036FO6SI', 'B000KL8ODE', '...

However, this metadata.json file takes more than 10 Go, for 9430088 entries (obtained by doing a wc -l metadata.json), so it does not fit in memory. Thus, as we will do a lot of tests later, we wanted to create a subfile containing only the Books metadata (we don't need video games metadata for example), with a subset of columns. We also want to write it in the csv format, to manipulate it in an easier way later.

We use the regex "\[\'books" in an ignore case mode, to obtain only entries that have a category tag beginning with [Books. In fact, if we want to use the regex 'book', some entries like 0078800242 or B00000078S are not books at all, even if there is Books in the title or the category tag. The '[' is useful here to avoir this behavior.

In [4]:
def read_book_metadata(path, regex): 
    g = open(path, 'r') 
    for l in g: 
        book = regex.search(l)
        if book:
            yield eval(l) 
            
def write_df_books_metadata(from_, to, regex, columns_to_keep): 
    i = 0 
    df = {} 
    for d in read_book_metadata(from_, regex): 
        df[i] = d 
        i += 1 
        if i % 10000 == 0: # Here, we choose to write everything every 10'000 book entries, and clear the dataframe to free memory.
            pd.DataFrame.from_dict(df, orient='index')[columns_to_keep].to_csv(to, header=False,mode='a')
            df = {}

COLUMNS_TO_KEEP = ['asin', 'salesRank', 'categories', 'title', 'price']
regex = re.compile('\[\'books', re.IGNORECASE)

if WRITE_BOOKS_METADATA:
    pd.DataFrame(columns=[COLUMNS_TO_KEEP]).to_csv(books_metadata_path)
    write_df_books_metadata(metadata_path, books_metadata_path, regex, COLUMNS_TO_KEEP)

And if we read what we just wrote :

In [5]:
read_csv(books_metadata_path)

Unnamed: 0     asin            salesRank   categories  \
0           0  1048791   {'Books': 6334800}  [['Books']]   
1           1  1048775  {'Books': 13243226}  [['Books']]   

                                               title  price  
0  The Crucible: Performed by Stuart Pankin, Jero...    NaN  
1     Measure for Measure: Complete &amp; Unabridged    NaN

We also wanted to obtain the ebooks titles, price etc..
For the category filter, we have to use the same trick as for the Books one : "\[\'Kindle". Please note that some book metadatas above are in fact kindle store metadatas, because the category can contain both. However, it's not a big deal if we want to do the merge later, because as we will see, pretty much no ebook has a title in the given metadata.

However, for the metadatas for ebooks, there was a problem at that step :

In [6]:
def read_ebook_metadata(path, regex): 
    g = open(path, 'r') 
    for l in g: 
        ebook = regex.search(l)
        if ebook:
            yield eval(l) 
def obtain_df_ebooks_metadata(from_, to, regex): 
    i = 0 
    df = {} 
    count = 0
    for d in read_ebook_metadata(from_, regex): 
        count += 1
        if(d.get('title')):
            df[i] = {'asin': d.get('asin'), 'title': d.get('title')}
            i += 1 
    pd.DataFrame.from_dict(df, orient='index').to_csv(to)
    print('Total ebooks in metadatas:', count)


regex = re.compile('\[\'Kindle', re.IGNORECASE)

if WRITE_EBOOKS_METADATA_TITLE:
    obtain_df_ebooks_metadata(metadata_path, ebooks_metadata_title_path, regex)

We see right below that 44 entries out of 434702 have a title. Of course, it's not good at all, since we want to merge books and ebooks using the title field.

In [7]:
read_csv(ebooks_metadata_title_path, None).shape[0]

44

Thus, we need to obtain the title field from another source. The first idea was to retrieve this information from Amazon directly, as we wanted to do for the user location. For that, we need to have a list of the ebooks asin (Amazon Standard Identification Numbers). We obtain it from the Kindle Store 5-core file.

In [8]:
def read_ebook_5core(path, regex): 
    g = open(path, 'r') 
    for l in g: 
        yield eval(l) 
def write_ebook_asin(from_, to): 
    i = 0 
    df = {} 
    for d in read_ebook_5core(from_, regex): 
        df[i] = d 
        i += 1 
        if i % 10000 == 0:
            if i % 100000 == 0:
                print(i) #to show the progression
            pd.DataFrame.from_dict(df, orient='index')[['asin']].to_csv(to, header=False,mode='a')
            df = {}
    pd.DataFrame.from_dict(df, orient='index')[['asin']].to_csv(to, header=False,mode='a')
    df = {}

if WRITE_EBOOK_ASIN:
    pd.DataFrame(columns=[['asin']]).to_csv(ebooks_asin)
    write_ebook_asin(kindle_5core_path, ebooks_asin)

As we were using the Kindle Store 5-core file, there are asin duplicates. We thus make it unique when we read.

In [3]:
ebooks_asin_unique = pd.read_csv('ebooks_asin.csv',usecols=[1]).asin.unique()
ebooks_asin_unique

array(['B000F83SZQ', 'B000FA64PA', 'B000FA64PK', ..., 'B00M029T4O',
       'B00M0RE7CS', 'B00M13FNSS'], dtype=object)

For every Amazon article with asin *xasinx*, the corresponding web page is https://www.amazon.com/dp/*xasinx*/ref=rdr_kindle_ext_tmb.


In [10]:
prefix = 'https://www.amazon.com/dp/'
suffix = '/ref=rdr_kindle_ext_tmb'

USER_AGENT_CHOICES = [
    'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:23.0) Gecko/20100101 Firefox/23.0',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/29.0.1547.62 Safari/537.36',
    'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; WOW64; Trident/6.0)',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/33.0.1750.146 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/33.0.1750.146 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64; rv:24.0) Gecko/20140205 Firefox/24.0 Iceweasel/24.3.0',
    'Mozilla/5.0 (Windows NT 6.2; WOW64; rv:28.0) Gecko/20100101 Firefox/28.0',
    'Mozilla/5.0 (Windows NT 6.2; WOW64; rv:28.0) AppleWebKit/534.57.2 (KHTML, like Gecko) Version/5.1.7 Safari/534.57.2',
]

We have to rotate the user-agent so that the bot is less likely to be considered as one. This is why we have a User-Agent array.
At the beginning, this method was working quite well : we had obtained 503 tuples (title, category, page number, language) over 1000 requests, which could mean that we had solved the problem. However, when looking at the distribution of the 503 tuples, we could see that at the beginning, everything behaves well, we obtain most of the entries (the other ones being like the B000JMKU0Y one, an obsolete entry, that only has customer reviews).

In [11]:
if AMAZON_GET_TITLE:
    
    LIMIT = 10
    
    undefined = 0
    defined = 0
    dataframe_original = pd.DataFrame(columns=[['asin', 'title', 'Category', 'PageNum', 'Language']])
    dataframe = dataframe_original.copy()

    dataframe_original.to_csv(amazon_ebooks)

    for i, asin in enumerate(ebooks_asin_unique[:LIMIT]):

        if i%10==0:
            headers = {'User-Agent':USER_AGENT_CHOICES[np.random.randint(0, len(USER_AGENT_CHOICES))]}
            if i > 0:
                print('undefined:', undefined, '/ defined:', defined)
                dataframe.to_csv(amazon_ebooks, header=False,mode='a')
                dataframe = dataframe_original.copy()


        r = requests.get(prefix + asin + suffix, headers=headers)
        page_body = r.text
        soup = BeautifulSoup(page_body, 'html.parser')
        title = soup.find_all('span', id='ebooksProductTitle')
        if(len(title) == 0):
            undefined += 1
        else:
            defined += 1
            title = title[0].text

            ul = soup.find_all('ul', class_='a-unordered-list a-horizontal a-size-small')
            if(len(ul) > 0):
                details = ul[0].find_all('a', class_='a-link-normal a-color-tertiary')
                if(len(details) > 0):
                    category = details[-1].text.strip()
                else:
                    category = ""
            else:
                category = ""

            details = soup.find_all('table', id='productDetailsTable')
            if(len(details) > 0):
                length = details[0].find_all('b', text='Print Length:')
                if(len(length) > 0):
                    page_number = length[0].parent.text.split()[2]
                else:
                    page_number = 0

                length = details[0].find_all('b', text='Language:')
                if(len(length) > 0):
                    language = length[0].parent.text.split()[1]
                else:
                    language = ""
            else:
                page_number = pd.np.nan
                language = ""

            dataframe.loc[asin] = (asin, title, category, page_number, language)

        waiting = np.random.rand()
        time.sleep(waiting+1)

    print(defined,',',undefined)
    dataframe.to_csv(amazon_ebooks, header=False,mode='a')
    dataframe = dataframe_original.copy()

But after some time, we get less and less entries : a message is sent by Amazon when retrieving the page, saying that it's not a good idea to continue scraping data, and that it might be a good idea to go through their API. So, there were some options :
- we continue to work with the bot while tweaking the parameters to behave like a normal user for the bot (by increasing the waiting time and rotating the user-agent as said before) :

After some online search (https://www.scrapehero.com/how-to-prevent-getting-blacklisted-while-scraping/,
http://blog.datahut.co/tutorial-how-to-scrape-amazon-using-python-scrapy/,
http://docs.aws.amazon.com/AWSECommerceService/latest/DG/rest-signature.html,
https://www.scrapehero.com/tutorial-how-to-scrape-amazon-product-details-using-python/,
https://blog.hartleybrody.com/scrape-amazon/), we saw that Amazon was detecting the IP, it could ban it, and the solution to avoid it was to use a proxy crawler. As it costs money, we decided not to use that. Furthermore, as said here, it's a legally speaking grey area : https://benbernardblog.com/web-scraping-and-crawling-are-perfectly-legal-right/.

- we try to go through the Amazon API : 

For the standard account, we need to put bank account information, so we prefer not to do so. For the student account, as we realized some days before the deadline that it existed, we might consider this option in the future if needed, but we wait for the epfl to accept or not the account request.

- we find a field in the metadata, different from the title, that can help us to merge a book with an ebook :

With some manual analysis, we found a pair of book-ebook : 

    {'asin': 'B000JML1QG', 'price': 0.99, 'imUrl': 'http://ecx.images-amazon.com/images/I/41VbZ%2BvxslL._BO2,204,203,200_PIsitb-sticker-v3-big,TopRight,0,-55_SX278_SY278_PIkin4,BottomRight,1,22_AA300_SH20_OU01_.jpg', 'related': {'also_viewed': ['B005LSCQ4Y', 'B0082UXYTE', 'B004TS2B4W'], 'buy_after_viewing': ['B00CS6P31U', 'B005LSCQ4Y', 'B0051EZX8Y', 'B006CRC98G']}, 'categories': [['Books', "Children's Books", 'Fairy Tales, Folk Tales & Myths', 'Anthologies'], ['Books', 'Literature & Fiction'], ['Kindle Store', 'Kindle eBooks', "Children's eBooks", 'Fairy Tales, Folk Tales & Myths', 'Anthologies'], ['Kindle Store', 'Kindle eBooks', "Children's eBooks", 'Fairy Tales, Folk Tales & Myths', 'Collections'], ['Kindle Store', 'Kindle eBooks', 'Literature & Fiction', 'Mythology & Folk Tales'], ['Kindle Store', 'Kindle eBooks', 'Science Fiction & Fantasy', 'Fantasy', 'Fairy Tales']]}


    {'asin': '0554319187', 'title': "Grimm's Fairy Stories", 'price': 0.99, 'imUrl': 'http://ecx.images-amazon.com/images/I/41O2olixwXL.jpg', 'related': {'also_viewed': ['1607103133', '0394709306', '1937994317'], 'buy_after_viewing': ['1607103133', '0394709306', '0393088863', '0385189508']}, 'salesRank': {'Books': 2586251}, 'categories': [['Books']]}

As we can see here, the only entry that is the same is the price, and it's dangerous to merge on the price as ebooks are often less expensive than the book version for the same content.

- we find another service that can give us the title for a given asin :

This is the option that we finally chose. The website http://asindb.com/ does exactly that. For this website, there is no bot detection as Amazon does. We can't retrieve the price, the category and the number of pages, but at least we can get the title. We can see the result below :

In [12]:
if ASINDB_GET_TITLE:
    
    LIMIT = 20
    
    prefix_asindb = 'http://asindb.com/USA/ASIN/'

    dataframe = pd.DataFrame(columns=[['asin', 'title']])
    notdefined = pd.DataFrame(columns=[['asin','notfound']])
    dataframe.to_csv(asindb_ebooks)

    for i, asin in enumerate(ebooks_asin_unique[:LIMIT]):
        r = requests.get(prefix_asindb + asin, headers={'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:23.0) Gecko/20100101 Firefox/23.0'})
        page_body = r.text
        soup = BeautifulSoup(page_body, 'html.parser')

        if i%10 == 0:
            dataframe.to_csv(asindb_ebooks, header=False,mode='a')
            dataframe = pd.DataFrame(columns=[['asin', 'title']])

        notfound = soup.find_all('h6', text = 'No item found!!!')
        if(len(notfound) > 0):
            notdefined.loc[asin] = (asin,1)
        else:
            title = soup.find_all('th', text='Title')
            if(len(title) > 0 and len(title[0].parent.findChildren()) >= 2):
                dataframe.loc[asin] = (asin, title[0].parent.findChildren()[1].text)
            else:
                print('alerte :', asin)

        waiting = np.random.rand()
        time.sleep(waiting+1)

    dataframe.to_csv(asindb_ebooks, header=False,mode='a')
    dataframe = pd.DataFrame(columns=[['asin', 'title']])
    
    print(notdefined.head(2))

Right above is the undefined entries dataframe (see below for more explanation).

We can see below what kind of output it gives to us.

In [13]:
ebooks_metadata = read_csv(asindb_ebooks, None)
ebooks_metadata.head()

Unnamed: 0        asin                                              title
0  B000FA64PA  B000FA64PA  Saboteur: Star Wars Legends (Darth Maul) (Shor...
1  B000FA64PK  B000FA64PK  Recovery: Star Wars Legends (The New Jedi Orde...
2  B000FA64QO  B000FA64QO  Ylesia: Star Wars Legends (The New Jedi Order)...
3  B000FBFMVG  B000FBFMVG  A Forest Apart: Star Wars Legends (Short Story...
4  B000FC1BN8  B000FC1BN8  Fool's Bargain: Star Wars Legends (Novella) (S...

In [14]:
read_csv(asindb_ebooks, None).shape[0] # number of titles retrieved

2741

This solution is of course not the best one : the asindb website does not contain everything. We have managed to retrieve 2741 titles over 4000 asins by using this technique, but we have no problem with the Amazon bot detection (and possible ban). We can see which entries were not retrieved by printing the notdefined dataframe.

Of course, tu retrieve the 2741 entries, we set the LIMIT constant in the code to be 4000.

We thus continue our analysis by using it.

Now, as we have the title information for books and ebooks, let's merge them. We read the book metadata information in books_metadata and we have the ebook metadata information with title in ebooks_metadata.

In [15]:
books_metadata = pd.read_csv(books_metadata_path)
books_metadata.head()

Unnamed: 0        asin            salesRank   categories  \
0           0  0001048791   {'Books': 6334800}  [['Books']]   
1           1  0001048775  {'Books': 13243226}  [['Books']]   
2           2  0001048236   {'Books': 8973864}  [['Books']]   
3           3  0000401048   {'Books': 6448843}  [['Books']]   
4           4  0001019880   {'Books': 9589258}  [['Books']]   

                                               title  price  
0  The Crucible: Performed by Stuart Pankin, Jero...    NaN  
1     Measure for Measure: Complete &amp; Unabridged    NaN  
2               The Sherlock Holmes Audio Collection   9.26  
3  The rogue of publishers' row;: Confessions of ...    NaN  
4          Classic Soul Winner's New Testament Bible   5.39

We merge them on title, and we see that we only get 1297 entries. It's not bad, but we can for sure have a better result. As we can remark, there is a lot of time the title The Space Between. We will discuss later about it.

In [16]:
#We merge, and get the head. Everything after is here just to have a nicer representation of the first elements
books_metadata.merge(ebooks_metadata, left_on='title', right_on='title').head().iloc[:,[1,7,2,3,4,5]].rename(columns={'asin_x':'asin_books', 'asin_y':'asin_ebooks'})

asin_books asin_ebooks           salesRank   categories              title  \
0  0002008505  B002DYJ7DM  {'Books': 5587764}  [['Books']]  The Space Between   
1  0615891411  B002DYJ7DM  {'Books': 4145053}  [['Books']]  The Space Between   
2  1579660584  B002DYJ7DM  {'Books': 2486827}  [['Books']]  The Space Between   
3  1588515508  B002DYJ7DM  {'Books': 5985767}  [['Books']]  The Space Between   
4  1595143394  B002DYJ7DM   {'Books': 141638}  [['Books']]  The Space Between   

   price  
0   4.74  
1   2.99  
2  20.08  
3    NaN  
4   7.13

In [17]:
books_metadata.merge(ebooks_metadata, left_on='title', right_on='title').shape[0]

1297

As we were saying, 1297 is not a big number, and we can do better. We have done the most basic possible thing to do : we have put the title for books and ebooks in lower form (minuscule).

In [18]:
books_metadata['title'] = books_metadata.title.str.lower()
books_metadata.head()

Unnamed: 0        asin            salesRank   categories  \
0           0  0001048791   {'Books': 6334800}  [['Books']]   
1           1  0001048775  {'Books': 13243226}  [['Books']]   
2           2  0001048236   {'Books': 8973864}  [['Books']]   
3           3  0000401048   {'Books': 6448843}  [['Books']]   
4           4  0001019880   {'Books': 9589258}  [['Books']]   

                                               title  price  
0  the crucible: performed by stuart pankin, jero...    NaN  
1     measure for measure: complete &amp; unabridged    NaN  
2               the sherlock holmes audio collection   9.26  
3  the rogue of publishers' row;: confessions of ...    NaN  
4          classic soul winner's new testament bible   5.39

In [19]:
ebooks_metadata['title'] = ebooks_metadata.title.str.lower()
ebooks_metadata.head()

Unnamed: 0        asin                                              title
0  B000FA64PA  B000FA64PA  saboteur: star wars legends (darth maul) (shor...
1  B000FA64PK  B000FA64PK  recovery: star wars legends (the new jedi orde...
2  B000FA64QO  B000FA64QO  ylesia: star wars legends (the new jedi order)...
3  B000FBFMVG  B000FBFMVG  a forest apart: star wars legends (short story...
4  B000FC1BN8  B000FC1BN8  fool's bargain: star wars legends (novella) (s...

In [20]:
merged_metadatas = books_metadata.merge(ebooks_metadata, left_on='title', right_on='title')
merged_metadatas = merged_metadatas[['asin_x', 'asin_y', 'title', 'price', 'categories', 'salesRank']]
merged_metadatas = merged_metadatas.rename(columns={'asin_x': 'asin_book', 'asin_y': 'asin_ebook'})

In [21]:
merged_metadatas.head()

asin_book  asin_ebook              title  price   categories  \
0  0002008505  B002DYJ7DM  the space between   4.74  [['Books']]   
1  0615891411  B002DYJ7DM  the space between   2.99  [['Books']]   
2  1579660584  B002DYJ7DM  the space between  20.08  [['Books']]   
3  1588515508  B002DYJ7DM  the space between    NaN  [['Books']]   
4  1595143394  B002DYJ7DM  the space between   7.13  [['Books']]   

            salesRank  
0  {'Books': 5587764}  
1  {'Books': 4145053}  
2  {'Books': 2486827}  
3  {'Books': 5985767}  
4   {'Books': 141638}

As we see, we have a bigger number of entries. We could have tried to increase this number, however as we will see later, we already have some problems with this 'strict' way of doing.

In [22]:
merged_metadatas.shape[0]

1506

So we have merged the two dataframes into one, and it seems that we can do some analysis on it. We have 1506 entries, so it's good for a first analysis in milestone 2 to do so.
But, as said before, there are title duplicates. It corresponds to books (there is only one duplicate entry for ebooks, for the article 'Second Chances') that have the same title.

In [23]:
merged_metadatas.head(7)

asin_book  asin_ebook              title  price   categories  \
0  0002008505  B002DYJ7DM  the space between   4.74  [['Books']]   
1  0615891411  B002DYJ7DM  the space between   2.99  [['Books']]   
2  1579660584  B002DYJ7DM  the space between  20.08  [['Books']]   
3  1588515508  B002DYJ7DM  the space between    NaN  [['Books']]   
4  1595143394  B002DYJ7DM  the space between   7.13  [['Books']]   
5  1601540817  B002DYJ7DM  the space between  10.99  [['Books']]   
6  1625530226  B002DYJ7DM  the space between   6.99  [['Books']]   

             salesRank  
0   {'Books': 5587764}  
1   {'Books': 4145053}  
2   {'Books': 2486827}  
3   {'Books': 5985767}  
4    {'Books': 141638}  
5  {'Books': 11114107}  
6   {'Books': 2678708}

In [24]:
ebooks_metadata.title.describe()

count           2741
unique          2739
top       armageddon
freq               2
Name: title, dtype: object

We thus thought that we could consider only pairs that have a unique title in the whole dataframe. In that way, we only have articles that have a unique name, at least in the period in which the dataset has been created, so that we could only have exactly the same content for the book and the ebook. 

We thus drop all elements that have a title that exists more than one time in the dataframe.

It's naive, as we don't have all books and ebooks of Amazon, even for the period given, but we wanted to try.

In [25]:
merged_metadatas = merged_metadatas.drop_duplicates('title',keep=False)
merged_metadatas.head(7)

asin_book  asin_ebook  \
39   0007269854  B003ZUY19I   
95   0060517689  B0036ZAHDG   
121  0060595620  B00480P58K   
173  0060813032  B0049H8X86   
200  0061084220  B004QS98KU   
460  0140249249  B003XVYGXK   
461  0140259678  B003C1R5CA   

                                                 title  price   categories  \
39                                    the ice princess   7.59  [['Books']]   
95                                         in the mood   2.99  [['Books']]   
121                                 the sweetest taboo   8.70  [['Books']]   
173  dragons from the sea (the strongbow saga, book 2)   3.60  [['Books']]   
200                                      raven's bride   7.69  [['Books']]   
460                                               iced    NaN  [['Books']]   
461                                     a timely death    NaN  [['Books']]   

              salesRank  
39   {'Books': 2527081}  
95   {'Books': 2663548}  
121  {'Books': 2956119}  
173  {'Books': 1269204}  
200  {'Books': 2911109}  
460  {'Books': 1225702}  
461  {'Books': 3077680}

We now only have 148 elements in the merged collection. We reming the reader about the fact that we tried to obtain 4000 ebooks titles, we got only 2741. By merging, we got 1506 entries, and when we drop all elements that have duplicates, we only have 148 elements.

In [26]:
merged_metadatas.shape[0]

148

After some manual analysis, we have unfortunately seen that for most entries, even if there is only one tuple (asin_book, asin_ebook), the two are not on the same content. We have an example for the first entry (the ice princess) : 'https://www.amazon.com/Princess-Patrik-Hedstrom-Erica-Falck/dp/0007269854' and 'https://www.amazon.com/Ice-Princess-Elizabeth-Hoyt-ebook/dp/B003ZUY19I' have the same title but are written by two different people, and have different content.

There are some tuples that match : for the article 'dragons from the sea (the strongbow saga, book 2)', we have the same content.

It's hard for us to quantify the number of such articles that do not match. We have done some by hand, and we have seen that a lot was not matching at all. A way to have good matches automatically could be to obtain the authors. We think that for the same author, it's rare to have two book with the same name. If we forget a minute about problems like the number of authors which is different for the ebook and the book even if it's the same, or different naming conventions (A fictive example : J. F. Brown or J. Brown), we would need the author information for each book and ebook of the tuples that we have merged.

The website we were <a href='http://asindb.com/'>using</a> does not provide this information. We then need to obtain it from somewhere else. 

We have tried to look at the library genesis, using this <a href='http://garbage.world/posts/libgen/'>tutorial</a>. However, there are only two ways to get back information about a book : using a special id (Libgen id), or by date. Furthermore, the asin field exists, but after some tests for which none of the articles had asin, it's hard to say if it's a good option. Thus, we have thought it's not the best solution for us.

We have also been told to look at the Gutenberg project. It could have been a good idea if we could search by asin in the metadatas, that are available. However, it seems that the asin data is not available, so we won't use it neither.

In [16]:
ebooks_asin_unique

array(['B000F83SZQ', 'B000FA64PA', 'B000FA64PK', ..., 'B00M029T4O',
       'B00M0RE7CS', 'B00M13FNSS'], dtype=object)

In [13]:
import bottlenose

In [14]:
(AWS_ACCESS_KEY_ID,AWS_SECRET_ACCESS_KEY,AWS_ASSOCIATE_TAG) = pd.read_csv('amazon_keys.csv', header=None, usecols=[1]).values.flatten()

In [15]:
def error_handler(err):
    ex = err['exception']
    if isinstance(ex, HTTPError) and ex.code == 503:
        time.sleep(random.expovariate(0.1))
        return True

In [16]:
amazon = bottlenose.Amazon(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_ASSOCIATE_TAG, ErrorHandler=error_handler)

In [18]:
keys = '0002178559'#,'.join(ebooks_asin_unique[10:20])

In [19]:
response = amazon.ItemLookup(ItemId=keys, ResponseGroup="Medium, Offers")
soup = BeautifulSoup(response, 'html.parser')
authors_retrieved = []
#print(soup.prettify())
items = soup.find_all('item')
for item in items[:10]:
    print(item.prettify())

    asin = item.find_all('asin')[0].text
    
    prices = item.find_all('price')
    for price in prices:
        print(price.find('amount').text)
    
    titles = item.find_all('title')
    if(len(titles) != 0):
        title_retrieved = titles[0].text.strip()

    manufacturers = item.find_all('manufacturer')
    if(len(manufacturers) != 0):
        manufacturer_retrieved = manufacturers[0].text.strip()

    authors = item.find_all('author')
    if(len(authors) != 0):
        for author in authors:
            authors_retrieved += [author.text.strip()]
    
    print(asin)
    print(title_retrieved)
    print(authors_retrieved)
    print(manufacturer_retrieved)
    print('--')

<item>
 <asin>
  0002178559
 </asin>
 <detailpageurl>
  https://www.amazon.com/Demanding-Impossible-Anarchism-Peter-Marshall/dp/0002178559?SubscriptionId=AKIAIQYRN6QJNHOGZS7Q&amp;tag=mj066-20&amp;linkCode=xm2&amp;camp=2025&amp;creative=165953&amp;creativeASIN=0002178559
 </detailpageurl>
 <itemlinks>
  <itemlink>
   <description>
    Technical Details
   </description>
   <url>
    https://www.amazon.com/Demanding-Impossible-Anarchism-Peter-Marshall/dp/tech-data/0002178559?SubscriptionId=AKIAIQYRN6QJNHOGZS7Q&amp;tag=mj066-20&amp;linkCode=xm2&amp;camp=2025&amp;creative=386001&amp;creativeASIN=0002178559
   </url>
  </itemlink>
  <itemlink>
   <description>
    Add To Baby Registry
   </description>
   <url>
    https://www.amazon.com/gp/registry/baby/add-item.html?asin.0=0002178559&amp;SubscriptionId=AKIAIQYRN6QJNHOGZS7Q&amp;tag=mj066-20&amp;linkCode=xm2&amp;camp=2025&amp;creative=386001&amp;creativeASIN=0002178559
   </url>
  </itemlink>
  <itemlink>
   <description>
    Add To Wedding

In [40]:
len(soup.find_all('itemattributes'))

6

In [19]:
BeautifulSoup(response, 'html.parser').prettify()

'<?xml version="1.0" ?>\n<itemlookupresponse xmlns="http://webservices.amazon.com/AWSECommerceService/2013-08-01">\n <operationrequest>\n  <httpheaders>\n   <header name="UserAgent" value="Python-urllib/3.6">\n   </header>\n  </httpheaders>\n  <requestid>\n   926a1e48-cba2-4483-8e34-d6fecf311abf\n  </requestid>\n  <arguments>\n   <argument name="AWSAccessKeyId" value="AKIAIQYRN6QJNHOGZS7Q">\n   </argument>\n   <argument name="AssociateTag" value="mj066-20">\n   </argument>\n   <argument name="ItemId" value="B000F83SZQ">\n   </argument>\n   <argument name="Operation" value="ItemLookup">\n   </argument>\n   <argument name="Service" value="AWSECommerceService">\n   </argument>\n   <argument name="Timestamp" value="2017-12-11T15:20:38Z">\n   </argument>\n   <argument name="Version" value="2013-08-01">\n   </argument>\n   <argument name="Signature" value="43RaHSog3IF2WPzYszFHJHLw7htznSRe/tVVdoKGGZ8=">\n   </argument>\n  </arguments>\n  <requestprocessingtime>\n   0.0801388650000000\n  </req

In [20]:
soup = BeautifulSoup(response, 'html.parser')

In [21]:
authors_retrieved = []


titles = soup.find_all('title')
if(len(titles) != 0):
    title_retrieved = titles[0].text.strip()

manufacturers = soup.find_all('manufacturer')
if(len(manufacturers) != 0):
    manufacturer_retrieved = manufacturers[0].text.strip()

authors = soup.find_all('author')
if(len(authors) != 0):
    for author in authors:
        authors_retrieved += [author.text.strip()]

print(title_retrieved)
print(authors_retrieved)
print(manufacturer_retrieved)

A Matter of Policy [An Amy Brewster Mystery]
['Sam Merwin Jr.']
PageTurner


In [36]:
responsebis = amazon.ItemLookup(ItemId=ebooks_asin_unique[1])

In [37]:
soup = BeautifulSoup(responsebis, 'html.parser')

In [38]:
soup.prettify()#.find('itemattributes')

'<?xml version="1.0" ?>\n<itemlookupresponse xmlns="http://webservices.amazon.com/AWSECommerceService/2013-08-01">\n <operationrequest>\n  <httpheaders>\n   <header name="UserAgent" value="Python-urllib/3.6">\n   </header>\n  </httpheaders>\n  <requestid>\n   35c7eb60-19f8-4ef6-ad82-579a10537759\n  </requestid>\n  <arguments>\n   <argument name="AWSAccessKeyId" value="AKIAIQYRN6QJNHOGZS7Q">\n   </argument>\n   <argument name="AssociateTag" value="mj066-20">\n   </argument>\n   <argument name="ItemId" value="B000FA64PA">\n   </argument>\n   <argument name="Operation" value="ItemLookup">\n   </argument>\n   <argument name="Service" value="AWSECommerceService">\n   </argument>\n   <argument name="Timestamp" value="2017-12-04T22:11:03Z">\n   </argument>\n   <argument name="Version" value="2013-08-01">\n   </argument>\n   <argument name="Signature" value="04tXLiapiuPH0/CiGX22iPOkm2Ui/C+MEmRWCfPjIAc=">\n   </argument>\n  </arguments>\n  <requestprocessingtime>\n   0.0793014240000000\n  </req

In [39]:
authors_retrieved = []


titles = soup.find_all('title')
if(len(titles) != 0):
    title_retrieved = titles[0].text.strip()

manufacturers = soup.find_all('manufacturer')
if(len(manufacturers) != 0):
    manufacturer_retrieved = manufacturers[0].text.strip()

authors = soup.find_all('author')
if(len(authors) != 0):
    for author in authors:
        authors_retrieved += [author.text.strip()]

print(title_retrieved)
print(authors_retrieved)
print(manufacturer_retrieved)

Saboteur: Star Wars Legends (Darth Maul) (Short Story) (Star Wars: Darth Maul Book 1)
['James Luceno', 'Ron Marz', 'Jan Duursema', 'Rick Magyar', 'Drew Struzan']
Del Rey


In [40]:
[ebooks_asin_unique[0], title_retrieved, 
                                          ['James Luceno', 'Ron Marz', 'Jan Duursema', 'Rick Magyar', 'Drew Struzan'], manufacturer_retrieved]

['B000F83SZQ',
 'Saboteur: Star Wars Legends (Darth Maul) (Short Story) (Star Wars: Darth Maul Book 1)',
 ['James Luceno', 'Ron Marz', 'Jan Duursema', 'Rick Magyar', 'Drew Struzan'],
 'Del Rey']

In [41]:
pd.DataFrame({'title':title_retrieved,'authors':", ".join(['James Luceno', 'Ron Marz', 'Jan Duursema', 'Rick Magyar', 'Drew Struzan']), 'manufacturer':manufacturer_retrieved},index=[ebooks_asin_unique[0]])

authors manufacturer  \
B000F83SZQ  James Luceno, Ron Marz, Jan Duursema, Rick Mag...      Del Rey   

                                                        title  
B000F83SZQ  Saboteur: Star Wars Legends (Darth Maul) (Shor...

In [42]:
dataframe_ebooks = pd.DataFrame(columns=['title', 'authors', 'manufacturer'])

In [43]:
new_entry = pd.DataFrame({'title':title_retrieved,'authors':", ".join(['James Luceno', 'Ron Marz', 'Jan Duursema', 'Rick Magyar', 'Drew Struzan']), 'manufacturer':manufacturer_retrieved},index=[ebooks_asin_unique[0]])
dataframe_ebooks = dataframe_ebooks.append(new_entry)
#dataframe_ebooks.append({'asin':ebooks_asin_unique[0], 'title':title_retrieved, 
#                                          'authors':authors_retrieved, 'manufacturer':manufacturer_retrieved}, ignore_index=True)
dataframe_ebooks[['title', 'authors', 'manufacturer']]

title  \
B000F83SZQ  Saboteur: Star Wars Legends (Darth Maul) (Shor...   

                                                      authors manufacturer  
B000F83SZQ  James Luceno, Ron Marz, Jan Duursema, Rick Mag...      Del Rey

In [44]:
#dataframe_undefined = pd.Series()

#dataframe_undefined.to_csv(amazon_aws_ebooks_undefined)

#dataframe_undefined = dataframe_undefined.append(pd.Series(ebooks_asin_unique[0]))
#dataframe_undefined = dataframe_undefined.append(pd.Series(ebooks_asin_unique[1]))

#dataframe_undefined.to_csv(amazon_aws_ebooks_undefined, header=False,mode='a')

In [45]:

new_entry = pd.DataFrame({'title':'','authors':", ".join([]), 'manufacturer':''},index=['111'])
dataframe_ebooks = dataframe_ebooks.append(new_entry)
dataframe_ebooks[['title', 'authors', 'manufacturer']].dropna(thresh=1, subset=[], inplace=False)

Empty DataFrame
Columns: [title, authors, manufacturer]
Index: []

In [46]:
log = pd.Series()
log = log.append(pd.Series(BeautifulSoup(responsebis, 'html.parser').prettify()))

In [47]:
log

0    <?xml version="1.0" ?>\n<itemlookupresponse xm...
dtype: object

In [48]:
#dataframe_undefined = pd.DataFrame(columns=['title', 'authors', 'manufacturer'])
#dataframe_undefined.to_csv(amazon_aws_ebooks_undefined)

In [49]:
import bottlenose

In [5]:
(AWS_ACCESS_KEY_ID,AWS_SECRET_ACCESS_KEY,AWS_ASSOCIATE_TAG) = pd.read_csv('amazon_keys.csv', header=None, usecols=[1]).values.flatten()

In [6]:
import random
import time
from urllib.error import HTTPError

def error_handler(err):
    ex = err['exception']
    if (isinstance(ex, HTTPError) and ex.code == 503) or isinstance(ex, TimeoutError):
        time.sleep(random.expovariate(0.1))
        return True

In [7]:
amazon = bottlenose.Amazon(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_ASSOCIATE_TAG, ErrorHandler=error_handler)

In [8]:
BEGIN = len(ebooks_asin_unique)
END = len(ebooks_asin_unique)
SIMULTANEOUS_REQUESTS = 10
STEP = 50
override = False

There are entries that are still not available through the Advertising API : B000FC26RI

Also newspaper don't comply to this contract of authors : The Wall Street Journal (B000FDJ0FS)

In [9]:
dataframe_ebooks = pd.DataFrame(columns=['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language'])
dataframe_undefined = pd.DataFrame(columns=['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language'])
log = pd.Series()

if(override):
    dataframe_ebooks.to_csv(amazon_aws_ebooks)
    dataframe_undefined.to_csv(amazon_aws_ebooks_undefined)
    log.to_csv(amazon_aws_ebooks_log)

ebooks_simultaneous_list = []

for i, ebook_asin in enumerate(ebooks_asin_unique[BEGIN:END]):
    ebooks_simultaneous_list += [ebook_asin]
    
    if i>0 and i%SIMULTANEOUS_REQUESTS == 0:
        keys = ','.join(ebooks_simultaneous_list)
        ebooks_simultaneous_list = []
        if i%STEP==0 and i > 0:
            print(i+BEGIN)
            dataframe_ebooks[['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language']].to_csv(amazon_aws_ebooks, header=False,mode='a')
            dataframe_ebooks = pd.DataFrame(columns=['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language'])

            dataframe_undefined[['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language']].to_csv(amazon_aws_ebooks_undefined, header=False,mode='a')
            dataframe_undefined = pd.DataFrame(columns=['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language'])

            log.to_csv(amazon_aws_ebooks_log, header=False,mode='a')
            log = pd.Series()

        response = amazon.ItemLookup(ItemId=keys, ResponseGroup="Medium, Offers")
        soup = BeautifulSoup(response, 'html.parser')
        
        items = soup.find_all('item')
        for item in items:
            all_important_defined = True
            authors_retrieved = []
            
            asins = item.find_all('asin')
            if(len(asins) != 0):
                asin_retrieved = asins[0].text.strip()
            else:
                asin_retrieved = ''
                all_important_defined = False
                
            salesRanks = item.find_all('salesrank')
            if(len(salesRanks) != 0):
                salesRank_retrieved = salesRanks[0].text.strip()
            else:
                salesRank_retrieved = ''
                #all_important_defined = False
                
            offers = item.find_all('offer')
            price_got = False
            for offer in offers:
                conditions = offer.find_all('condition')
                if(len(conditions) > 0 and conditions[0].text == 'New'):
                    prices = offer.find_all('amount')
                    curr_codes = offer.find_all('currencycode')
                    if(len(prices) > 0 and len(curr_codes) > 0):
                        price_retrieved = prices[0].text
                        curr_retrieved = curr_codes[0].text
                        price_got = True
                       
            if(not price_got):
                price_retrieved = ''
                curr_retrieved = ''
                #all_important_defined = False
                
            
            releaseDates = item.find_all('releasedate')
            if(len(releaseDates) != 0):
                reldate_retrieved = releaseDates[0].text.strip()
            else:
                reldate_retrieved = ''
                all_important_defined = False
                
            pages = item.find_all('numberofpages')
            if(len(pages) != 0):
                page_retrieved = pages[0].text.strip()
            else:
                page_retrieved = ''
                #all_important_defined = False
            
            languages = item.find_all('language')
            if(len(languages) != 0):
                language_retrieved = languages[0].find('name').text.strip()
            else:
                language_retrieved = ''
                #all_important_defined = False
                

            titles = item.find_all('title')
            if(len(titles) != 0):
                title_retrieved = titles[0].text.strip()
            else:
                title_retrieved = ''
                all_important_defined = False

            manufacturers = item.find_all('manufacturer')
            if(len(manufacturers) != 0):
                manufacturer_retrieved = manufacturers[0].text.strip()
            else:
                manufacturer_retrieved = ''

            authors = item.find_all('author')
            if(len(authors) != 0):
                for author in authors:
                    authors_retrieved += [author.text.strip()]
            else:
                authors_retrieved = []
                all_important_defined = False

            new_entry = pd.DataFrame({'title':title_retrieved,'authors':", ".join(authors_retrieved), 'manufacturer':manufacturer_retrieved, 'rank': salesRank_retrieved, 'price':price_retrieved, 'currency':curr_retrieved, 'reldate':reldate_retrieved, 'page': page_retrieved, 'language': language_retrieved},index=[asin_retrieved])
            if(all_important_defined):
                dataframe_ebooks = dataframe_ebooks.append(new_entry)
            else:
                dataframe_undefined = dataframe_undefined.append(new_entry)
                log = log.append(pd.Series(item.prettify()))

dataframe_ebooks[['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language']].to_csv(amazon_aws_ebooks, header=False,mode='a')
dataframe_undefined[['title', 'authors', 'manufacturer', 'rank', 'price', 'currency', 'reldate', 'page', 'language']].to_csv(amazon_aws_ebooks_undefined, header=False,mode='a')
log.to_csv(amazon_aws_ebooks_log, header=False,mode='a')

1050
1100
1150
1200
1250
1300
1350
1400
1450
1500
1550
1600
1650
1700
1750
1800
1850
1900
1950
2000
2050
2100
2150
2200
2250
2300
2350
2400
2450
2500
2550
2600
2650
2700
2750
2800
2850
2900
2950
3000
3050
3100
3150
3200
3250
3300
3350
3400
3450
3500
3550
3600
3650
3700
3750
3800
3850
3900
3950
4000
4050
4100
4150
4200
4250
4300
4350
4400
4450
4500
4550
4600
4650
4700
4750
4800
4850
4900
4950
5000
5050
5100
5150
5200
5250
5300
5350
5400
5450
5500
5550
5600
5650
5700
5750
5800
5850
5900
5950
6000
6050
6100
6150
6200
6250
6300
6350
6400
6450
6500
6550
6600
6650
6700
6750
6800
6850
6900
6950
7000
7050
7100
7150
7200
7250
7300
7350
7400
7450
7500
7550
7600
7650
7700
7750
7800
7850
7900
7950
8000
8050
8100
8150
8200
8250
8300
8350
8400
8450
8500
8550
8600
8650
8700
8750
8800
8850
8900
8950
9000
9050
9100
9150
9200
9250
9300
9350
9400
9450
9500
9550
9600
9650
9700
9750
9800
9850
9900
9950
10000
10050
10100
10150
10200
10250
10300
10350
10400
10450
10500
10550
10600
10650
10700
10750
10800
108

In [91]:
amazon_ebooks_metadata = pd.read_csv(amazon_aws_ebooks).rename(columns={'Unnamed: 0':'asin'}).drop_duplicates(subset='asin')
amazon_ebooks_metadata['title'] = amazon_ebooks_metadata.title.str.lower()
amazon_ebooks_metadata['authors'] = amazon_ebooks_metadata.authors.str.lower()
#amazon_ebooks_metadata['title'] = amazon_ebooks_metadata['title'].str.replace(r"\(.*\)|\[.*\]","")

In [92]:
amazon_ebooks_metadata = amazon_ebooks_metadata.dropna(subset=['authors'])
amazon_ebooks_metadata_multiple_authors = amazon_ebooks_metadata[amazon_ebooks_metadata.authors.str.contains(',')]

In [93]:
authors = amazon_ebooks_metadata_multiple_authors.authors.str.split(',')
count = 0

new_df = pd.DataFrame(columns=['asin', 'title', 'authors', 'manufacturer'])

for i,(index, row) in enumerate(amazon_ebooks_metadata_multiple_authors.iterrows()):
    for author in authors.iloc[i]:
        new_df.loc[count] = row
        new_df.loc[count].authors = author
        count += 1
new_df

asin                                              title  \
0     B000FA64PA  saboteur: star wars legends (darth maul) (shor...   
1     B000FA64PA  saboteur: star wars legends (darth maul) (shor...   
2     B000FA64PA  saboteur: star wars legends (darth maul) (shor...   
3     B000FA64PA  saboteur: star wars legends (darth maul) (shor...   
4     B000FA64PA  saboteur: star wars legends (darth maul) (shor...   
5     B000FA64PK  recovery: star wars legends (the new jedi orde...   
6     B000FA64PK  recovery: star wars legends (the new jedi orde...   
7     B000FA64PK  recovery: star wars legends (the new jedi orde...   
8     B000FA64PK  recovery: star wars legends (the new jedi orde...   
9     B000FA64PK  recovery: star wars legends (the new jedi orde...   
10    B000PDZG6K                                          tallchief   
11    B000PDZG6K                                          tallchief   
12    B001AWYAZ8  the life of a child survivor from bialystok, p...   
13    B001AWYAZ8  the life of a child survivor from bialystok, p...   
14    B001AWYAZ8  the life of a child survivor from bialystok, p...   
15    B001EWEB5I  an autobiography: the story of my experiments ...   
16    B001EWEB5I  an autobiography: the story of my experiments ...   
17    B001TK41W8                         from the cell to the cross   
18    B001TK41W8                         from the cell to the cross   
19    B001VH6NWU  reality creation 101: mastering manifestation ...   
20    B001VH6NWU  reality creation 101: mastering manifestation ...   
21    B001VH6NWU  reality creation 101: mastering manifestation ...   
22    B001VH6NWU  reality creation 101: mastering manifestation ...   
23    B001VH6NWU  reality creation 101: mastering manifestation ...   
24    B00200JXAG                                  campaign for love   
25    B00200JXAG                                  campaign for love   
26    B00264H2VI  65 proof - jack daniels and other thriller sto...   
27    B00264H2VI  65 proof - jack daniels and other thriller sto...   
28    B0026ZOJ18        the christian novel collection (four books)   
29    B0026ZOJ18        the christian novel collection (four books)   
...          ...                                                ...   
6022  B00LQB1IIU                   body games (games series book 5)   
6023  B00LQTGPWG               inferno: a devil chaser's mc romance   
6024  B00LQTGPWG               inferno: a devil chaser's mc romance   
6025  B00LRSN11Y                       ricochet (renegades, book 3)   
6026  B00LRSN11Y                       ricochet (renegades, book 3)   
6027  B00LS153XO  summer in new york collection (a timeless roma...   
6028  B00LS153XO  summer in new york collection (a timeless roma...   
6029  B00LS153XO  summer in new york collection (a timeless roma...   
6030  B00LS153XO  summer in new york collection (a timeless roma...   
6031  B00LS153XO  summer in new york collection (a timeless roma...   
6032  B00LS153XO  summer in new york collection (a timeless roma...   
6033  B00LS6MCMY     leather and lace (lonesome point texas book 1)   
6034  B00LS6MCMY     leather and lace (lonesome point texas book 1)   
6035  B00LT48ZR6  touched by the dark: a paranormal romance (imm...   
6036  B00LT48ZR6  touched by the dark: a paranormal romance (imm...   
6037  B00LTATG9G  diabetes cure: my against all odds story of ho...   
6038  B00LTATG9G  diabetes cure: my against all odds story of ho...   
6039  B00LTATG9G  diabetes cure: my against all odds story of ho...   
6040  B00LU6N5UU    into the shadows (undercover associates book 3)   
6041  B00LU6N5UU    into the shadows (undercover associates book 3)   
6042  B00LVBJOYK                                          hourglass   
6043  B00LVBJOYK                                          hourglass   
6044  B00LWRN8SQ                               the name on her lips   
6045  B00LWRN8SQ                               the name on her lips   
6046  B00LWSNUQ0                    

In [94]:
amazon_ebooks_metadata = pd.concat([amazon_ebooks_metadata, new_df])

In [95]:
amazon_ebooks_metadata

asin                                              title  \
0     B000F83SZQ       a matter of policy [an amy brewster mystery]   
1     B000FA64PA  saboteur: star wars legends (darth maul) (shor...   
2     B000FA64PK  recovery: star wars legends (the new jedi orde...   
3     B000FA64QO  ylesia: star wars legends (the new jedi order)...   
4     B000FBFMVG  a forest apart: star wars legends (short story...   
5     B000FC1BN8  fool's bargain: star wars legends (novella) (s...   
6     B000FC1TG2                                   the iron marshal   
7     B000FC2MB8                                   brionne: a novel   
8     B000JMKNQ0  boba fett: a practical man: star wars legends ...   
9     B000JMLBHU                              the mysterious island   
10    B000JMLG4S  knots, splices and rope work a practical treatise   
11    B000JMLHYC                                the english orphans   
12    B000JQUSBQ                                 indian fairy tales   
13    B000JQUT8S                                       oliver twist   
14    B000JQUTGK  the lincoln story book a judicious collection ...   
15    B000JQUU00                                english fairy tales   
16    B000JQUXBQ  the diary of a u-boat commanderwith an introdu...   
17    B000MAH6G4                       hawke: the law of a fast gun   
18    B000MAH73Q     falling in love again (avon romantic treasure)   
19    B000NY11RY                       tempt me, taste me, touch me   
20    B000PDZG6K                                          tallchief   
21    B000Q9J0CO  caught off guard [heroes of silver springs 1] ...   
22    B000QCS8YM  sherlock holmes: the complete novels and stori...   
23    B000QUCOB2          stealing sophie (avon historical romance)   
24    B000R3NNAE                           hawke: the king hill war   
25    B000SEGE3K                       sky hunters: anarchy's reign   
26    B000SN6IOQ                                   allan quatermain   
27    B000TU16PE                 the highland rogues: warrior bride   
28    B000U0NSV8          an unladylike offer (the radwells book 2)   
29    B000UB9NCK                                       the stranger   
...          ...                                                ...   
6022  B00LQB1IIU                   body games (games series book 5)   
6023  B00LQTGPWG               inferno: a devil chaser's mc romance   
6024  B00LQTGPWG               inferno: a devil chaser's mc romance   
6025  B00LRSN11Y                       ricochet (renegades, book 3)   
6026  B00LRSN11Y                       ricochet (renegades, book 3)   
6027  B00LS153XO  summer in new york collection (a timeless roma...   
6028  B00LS153XO  summer in new york collection (a timeless roma...   
6029  B00LS153XO  summer in new york collection (a timeless roma...   
6030  B00LS153XO  summer in new york collection (a timeless roma...   
6031  B00LS153XO  summer in new york collection (a timeless roma...   
6032  B00LS153XO  summer in new york collection (a timeless roma...   
6033  B00LS6MCMY     leather and lace (lonesome point texas book 1)   
6034  B00LS6MCMY     leather and lace (lonesome point texas book 1)   
6035  B00LT48ZR6  touched by the dark: a paranormal romance (imm...   
6036  B00LT48ZR6  touched by the dark: a paranormal romance (imm...   
6037  B00LTATG9G  diabetes cure: my against all odds story of ho...   
6038  B00LTATG9G  diabetes cure: my against all odds story of ho...   
6039  B00LTATG9G  diabetes cure: my against all odds story of ho...   
6040  B00LU6N5UU    into the shadows (undercover associates book 3)   
6041  B00LU6N5UU    into the shadows (undercover associates book 3)   
6042  B00LVBJOYK                                          hourglass   
6043  B00LVBJOYK                                          hourglass   
6044  B00LWRN8SQ                               the name on her lips   
6045  B00LWRN8SQ                               the name on her lips   
6046  B00LWSNUQ0                    

In [96]:
amazon_books_metadata_yves = pd.read_csv(amazon_aws_books_yves, encoding = "latin1").rename(columns={'Unnamed: 0':'asin'}).drop_duplicates(subset='asin')
amazon_books_metadata_yves['title'] = amazon_books_metadata_yves.title.str.lower()
amazon_books_metadata_yves['authors'] = amazon_books_metadata_yves.authors.str.lower()
#amazon_books_metadata_yves['title'] = amazon_books_metadata_yves['title'].str.replace(r"\(.*\)|\[.*\]","")
amazon_books_metadata_yves = amazon_books_metadata_yves.dropna(subset=['authors'])
amazon_books_metadata_yves_multiple_authors = amazon_books_metadata_yves[amazon_books_metadata_yves.authors.str.contains(',')]

authors = amazon_books_metadata_yves_multiple_authors.authors.str.split(',')
count = 0

new_df = pd.DataFrame(columns=['asin', 'title', 'authors', 'manufacturer'])

for i,(index, row) in enumerate(amazon_books_metadata_yves_multiple_authors.iterrows()):
    for author in authors.iloc[i]:
        new_df.loc[count] = row
        new_df.loc[count].authors = author
        count += 1
new_df

amazon_books_metadata_yves = pd.concat([amazon_books_metadata_yves, new_df])

merged_amazon = amazon_books_metadata_yves.merge(amazon_ebooks_metadata, left_on='title', right_on='title').rename(columns={'asin_x':'asin_books', 'asin_y':'asin_ebooks'})
merged_amazon_prime = merged_amazon[merged_amazon.authors_x == merged_amazon.authors_y]
merged_amazon_prime

asin_books                                              title  \
93    0060813032  dragons from the sea (the strongbow saga, book 2)   
221   0061659282                  untamed city: carnival of secrets   
380   0140287957                                   an ideal husband   
381   0140364749                                         lad: a dog   
396   0141196882                                            dracula   
510   0307266931                                      war and peace   
761   0345548957                     a wedding in provence: a novel   
907   0380486032                                            rebecca   
1038  0393096696                              the turn of the screw   
1106  0425147770                            whispers from the grave   
1191  0440134986                                  a heart too proud   
1240  0446363049                                             vortex   
1244  0446363049                                             vortex   
1248  0446363049                                             vortex   
1338  0451191137                                             anthem   
1345  0451202821                                her daughter's eyes   

                       authors_x       manufacturer_x asin_ebooks  \
93                judson roberts           HarperTeen  B0049H8X86   
221                 melissa marr           HarperColl  B007HBY0VK   
380                  oscar wilde        Penguin Books  B004UJCKQY   
381        albert payson terhune         Puffin Books  B0076QWBD0   
396                  bram stoker     Penguin Classics  B00J69Y3PY   
510                  leo tolstoy                Knopf  B003X4M9F4   
761                ellen sussman     Ballantine Books  B00I1ZG5HI   
907            daphne du maurier           Avon Books  B00CO7FLJM   
1038                 henry james  W W Norton & Co Inc  B004UJIUD6   
1106                 leslie rule              Berkley  B00C6QQBZO   
1191                laura london                 Dell  B00CO7GGSM   
1240  larry bond, patrick larkin   Warner Books, Inc.  B008OM2BHQ   
1244                  larry bond   Warner Books, Inc.  B008OM2BHQ   
1248              patrick larkin   Warner Books, Inc.  B008OM2BHQ   
1338                    ayn rand               Signet  B0082RHM4A   
1345    jessica barksdale inclan            NAL Trade  B0051ZN10A   

                       authors_y                  manufacturer_y  
93                judson roberts                  Northman Books  
221                 melissa marr                   HarperCollins  
380                  oscar wilde                             NaN  
381        albert payson terhune                             NaN  
396                  bram stoker                Simon & Schuster  
510                  leo tolstoy  Superior Formatting Publishing  
761                ellen sussman                Ballantine Books  
907            daphne du maurier       Little, Brown and Company  
1038                 henry james                             NaN  
1106                 leslie rule            Planet Ann Rule, LLC  
1191                laura london                   Forever Yours  
1240  larry bond, patrick larkin                             NaN  
1244                  larry bond                             NaN  
1248              patrick larkin                             NaN  
1338                    ayn rand                             NaN  
1345    jessica barksdale inclan                             NaN

In [97]:
merged_amazon_prime.shape

(16, 7)

In [98]:
amazon_books_metadata_pa = pd.read_csv(amazon_aws_books_pa).rename(columns={'Unnamed: 0':'asin'}).drop_duplicates(subset='asin')
amazon_books_metadata_pa['title'] = amazon_books_metadata_pa.title.str.lower()
amazon_books_metadata_pa['authors'] = amazon_books_metadata_pa.authors.str.lower()
#amazon_books_metadata_pa['title'] = amazon_books_metadata_pa['title'].str.replace(r"\(.*\)|\[.*\]","")
amazon_books_metadata_pa = amazon_books_metadata_pa.dropna(subset=['authors'])
amazon_books_metadata_pa_multiple_authors = amazon_books_metadata_pa[amazon_books_metadata_pa.authors.str.contains(',')]

authors = amazon_books_metadata_pa_multiple_authors.authors.str.split(',')
count = 0

new_df = pd.DataFrame(columns=['asin', 'title', 'authors', 'manufacturer'])

for i,(index, row) in enumerate(amazon_books_metadata_pa_multiple_authors.iterrows()):
    for author in authors.iloc[i]:
        new_df.loc[count] = row
        new_df.loc[count].authors = author
        count += 1
new_df

amazon_books_metadata_pa = pd.concat([amazon_books_metadata_pa, new_df])




merged_amazon = amazon_books_metadata_pa.merge(amazon_ebooks_metadata, left_on='title', right_on='title').rename(columns={'asin_x':'asin_books', 'asin_y':'asin_ebooks'})
merged_amazon_prime = merged_amazon[merged_amazon.authors_x == merged_amazon.authors_y]
merged_amazon_prime

asin_books                                 title               authors_x  \
89   0679642366                 the mysterious island             jules verne   
147  0735822271                                 heidi           johanna spyri   
337  0758242212                     sweet little lies           michele grant   
399  0760712727  benedict arnold: patriot and traitor  willard sterne randall   
525  0786010924                      all the way home       wendy corsi staub   
526  0786011327                              moonfall           tamara thorne   
663  0843946237                        beyond forever              debra dier   
680  084395261X                            open range            lauran paine   
688  0843959142                             star born            andre norton   
710  087004124X                                anthem                ayn rand   

            manufacturer_x asin_ebooks               authors_y  \
89          Modern Library  B000JMLBHU             jules verne   
147       NorthSouth Books  B004TPPYJY           johanna spyri   
337                 Dafina  B00DZURKXC           michele grant   
399           Dorset Press  B0083UL1F2  willard sterne randall   
525               Pinnacle  B009NQZY1C       wendy corsi staub   
526               Pinnacle  B009TZRUKK           tamara thorne   
663          Leisure Books  B00HDEEKJS              debra dier   
680  Dorchester Pub Co Inc  B009NXLYIM            lauran paine   
688                 Cosmos  B004TPG9ZM            andre norton   
710           Caxton Press  B0082RHM4A                ayn rand   

          manufacturer_y  
89   Public Domain Books  
147                  NaN  
337               Dafina  
399  New Word City, Inc.  
525               Harper  
526        Tamara Thorne  
663     Dierington Press  
680         AmazonEncore  
688                  NaN  
710                  NaN

In [99]:
merged_amazon_prime.shape

(10, 7)

In [84]:
#pd.merge(amazon_ebooks_metadata.title.str.lower(), books_metadata.title.str.lower())
books_metadata.merge(amazon_ebooks_metadata, left_on='title', right_on='title').rename(columns={'asin_x':'asin_books', 'asin_y':'asin_ebooks'})

Unnamed: 0  asin_books            salesRank   categories  \
0             145  0001981692  {'Books': 11831050}  [['Books']]   
1          451645  0671873164   {'Books': 1117160}  [['Books']]   
2          820435  088388027X   {'Books': 1863186}  [['Books']]   
3         1087185  1416927344    {'Books': 248278}  [['Books']]   
4         1309521  1558594051  {'Books': 10853889}  [['Books']]   
5         1421569  1590784553    {'Books': 993678}  [['Books']]   
6         1494164  1606591827   {'Books': 2569651}  [['Books']]   
7         1779659  B0002MJDKO   {'Books': 7990596}  [['Books']]   
8             172  0002008211   {'Books': 5219330}  [['Books']]   
9          126720  0263743608   {'Books': 5898534}  [['Books']]   
10         187182  033045157X   {'Books': 1071929}  [['Books']]   
11         597731  0778328015   {'Books': 1218056}  [['Books']]   
12         611314  0786010908   {'Books': 2167685}  [['Books']]   
13         611493  0786016841    {'Books': 602214}  [['Books']]   
14         900588  0956410170                  NaN  [['Books']]   
15         946070  0972716025   {'Books': 2838504}  [['Books']]   
16        1012514  0986556904   {'Books': 7501299}  [['Books']]   
17        1204528  1466225890  {'Books': 10108097}  [['Books']]   
18        1277102  1494469308                  NaN  [['Books']]   
19        1285883  1499311877   {'Books': 2162014}  [['Books']]   
20        1544369  1783336285   {'Books': 2696259}  [['Books']]   
21        1634483  1897299265   {'Books': 2507766}  [['Books']]   
22        1759072  938015495X   {'Books': 3599478}  [['Books']]   
23            270  0002008505   {'Books': 5587764}  [['Books']]   
24         426833  0615891411   {'Books': 4145053}  [['Books']]   
25        1380542  1579660584   {'Books': 2486827}  [['Books']]   
26        1412333  1588515508   {'Books': 5985767}  [['Books']]   
27        1448217  1595143394    {'Books': 141638}  [['Books']]   
28        1477494  1601540817  {'Books': 11114107}  [['Books']]   
29        1532498  1625530226   {'Books': 2678708}  [['Books']]   
...           ...         ...                  ...          ...   
13178     1846370  B000TFTIC2   {'Books': 8889086}  [['Books']]   
13179     1847316  B000U44BIS   {'Books': 1103192}  [['Books']]   
13180     1847708  B000UDPEYY                  NaN  [['Books']]   
13181     1872509  B001JCVIGS                  NaN  [['Books']]   
13182     1847860  B000UEYYE4  {'Books': 10273803}  [['Books']]   
13183     1852063  B000X00RNC   {'Books': 6990961}  [['Books']]   
13184     1856506  B0011MRGLC   {'Books': 7813520}  [['Books']]   
13185     1863372  B0018E04JE    {'Books': 384423}  [['Books']]   
13186     1872060  B001ISOBLC   {'Books': 6393218}  [['Books']]   
13187     1872898  B001JYTR32                  NaN  [['Books']]   
13188     1874342  B001LOMPP2   {'Books': 4414963}  [['Books']]   
13189     1878431  B001QXC4O0   {'Books': 2717429}  [['Books']]   
13190     1883816  B0024QTCDO                  NaN  [['Books']]   
13191     1884156  B00262OJUM   {'Books': 8475458}  [['Books']]   
13192     1884160  B00262O8ZI  {'Books': 10932845}  [['Books']]   
13193     1887443  B002AJGLO8   {'Books': 4014766}  [['Books']]   
13194     1887479  B002ALUAJS   {'Books': 3624486}  [['Books']]   
13195     1929330  B0043NZKKU   {'Books': 3831799}  [['Books']]   
13196     1893680  B002KYF3SS   {'Books': 2386450}  [['Books']]   
13197     1895438  B002OBTPXQ    {'Books': 491310}  [['Books']]   
13198     1897893  B002SR8KTG   {'Books': 6629752}  [['Books']]   
13199     1911315  B003F20PSM  {'Books': 10132076}  [['Books']]   
13200     1919643  B003UILUJ4    {'Books': 544107}  [['Books']]   
13201     1926161  B0040V4H9U   {'Books': 9739966}  [['Books']]   
13202     1974255  B0052VHMEE   {'Books': 4965839}  [['Books']]   
13203     1996530  B005LE4RMC   {'Books': 8580459}  [['Books']]   
13204     2024317  B006ONEE0E   {'Books': 3003064}  [['Books']]   
13205     2046520  B007EPSQW8                  NaN  [[

-- --
## Review analysis

Even if we have some problems that will need to be fixed for the milestone 3, we have continued in the analysis. Of course we had to test with wrong data, but we will ensure that in the future, we will make it work on correct data (one with same author and title for books and ebooks). The code is here to show that we have worked on other steps of the pipeline.

In [27]:
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer

/Users/MJ/anaconda/lib/python3.6/site-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


Here you will need to uncomment this and download vader_lexicon as described in the README.

In [28]:
#nltk.download_shell()

Next we are going to fetch the reviews of the books for which we have the ebooks.

As the file for the reviews of the books is quite big, we are going the read it into chunks.

In [29]:
def find_matched(list_asins, in_path, out_path):
    pd.DataFrame(columns=[['asin','overall', 'summary', 'reviewerID', 'helpful','reviewText', 'reviewerName']]).to_csv(out_path)
    
    for chunk in pd.read_json(in_path, lines=True, chunksize=50000):
        
        # look if some asins of the chunk match the ones in the given list
        filtered = chunk[chunk['asin'].isin(list_asins)].dropna(how='all')
        
        if len(filtered) > 0:
            # write the data if any
            filtered[['asin','overall', 'summary', 'reviewerID', 'helpful','reviewText', 'reviewerName']].to_csv(out_path,header=False, mode='a')


In [30]:
if WRITE_FIND_MATCHED:
    find_matched(merged_metadatas['asin_book'].values, books_5core_path, matched_books_path)
    find_matched(merged_metadatas['asin_ebook'].values, kindle_5core_path, matched_ebooks_path)

Now that we have our reviews it is time to analyse their content. To compute the score of a book, we use two approaches:

1) A weighted average of the stars taking into account the helpfulness of the review as described below.

Let $s_{i,j}$ be the $j$th rating of book $i$ and $n$ the number of ratings for this book. As other users can review a review by saying whether it is helpful, let $k_{i,j}$ be the number of the person who reviewed the review $s_{i,j}$ and let $u_{i,j}$ be the number of reviewer who found the review helpful among the $k_{i,j}$ reviewers. Then what we want to do is to give more weight to reviews which are helpful and less weight to reviews that the other users found useless. The maximum weight a review can have is 1 and the minimum is 0, then if a review has never been reviewed or it has been voted equally helpful and not helpful, its weight will be 0.5 which is neutral.
$$w_{i,j}=\cases{\frac{u_{i,j}}{k_{i,j}}, \text{if } k_{i,j} \ne 0 \\0.5, \text{if } k_{i,j}=0}$$

    
The weighted average is then:

$$S_i = \frac{\sum_{j=1}^{n} w_{i,j}s_{i,j}}{\sum_{j=1}^{n} w_{i,j}}$$

2) A weighted average of the sentiment's intensity in the review taking into account the helpfulness with the weight being derived similarly as above.

Here we are going to use the VADER (Valence Aware Dictionary sEntiment Reasoner) sentiment analyzer from the nltk package. VADER is based on lexicons of sentiment-related words and each words is rated as whether it is positive and negative, and how negative or positive it is. For example, the 'excellent' would be treated as more positive than 'good'.

The score Vader returns is between -1 and 1, 1 for a very positive review, -1 for a very negative review, and 0 if it is neutral.

Although Vader is not the most accurate tool and to analyse a piece of text it checks if any of the words in the text are present in the lexicon, therefore its accuracy depends on the coverage of the lexicons. It is the easiest approach we have for the moment as we cannot train a classifier since we don't have a proper training set.

In [31]:
sid = SentimentIntensityAnalyzer()

In [32]:
# compute the weight of a review
def weighted_helpful(x):
    # literal_eval will evaluate '[1 ,1]' as a list
    # we use literal_eval here since using astype(list) on the column was not working
    x = literal_eval(x)
    voters = int(x[0]) + int(x[1])
    return 0.5 if voters == 0 else int(x[0])/voters

In [33]:
# compute the weighted average of the two scores for a book
def weighted_scores(data):
    # get the weight
    data['weighted_help'] = (data['helpful'].astype(list)).apply(weighted_helpful)
    
    func = lambda x: sid.polarity_scores(x)['compound']
    
    data['sentiment_review'] = data['reviewText'].apply(func)
    data['sentiment_summary'] = data['summary'].apply(func)
    
    data['sentiment_average'] = (data['sentiment_review'] + data['sentiment_summary'])*0.5
    data['sentiment_rescaled'] = (data['sentiment_average'] + 1) * 2 + 1
    
    data['combined'] = (data['sentiment_rescaled'] + data['overall']) * 0.5
    
    # multiply the scores with the weight
    data['weighted_sentiment_review'] = data['weighted_help'] * data['sentiment_review']
    data['weighted_sentiment_summary'] = data['weighted_help'] * data['sentiment_summary']
    data['weighted_sentiment_average'] = data['weighted_help'] * data['sentiment_average']
    data['weighted_sentiment_rescaled'] = data['weighted_help'] * data['sentiment_rescaled']
    data['weighted_overall'] = data['weighted_help'] * data['overall']
    data['weighted_combined'] = data['weighted_help'] * data['combined']
    
    # sum everything
    weighted_score = data.groupby(data['asin']).sum()
    
    # divide by the sum of the weights to obtain the weighted average
    weighted_score['sentiment_review_score'] = weighted_score['weighted_sentiment_review']/weighted_score['weighted_help']
    weighted_score['sentiment_summary_score'] = weighted_score['weighted_sentiment_summary']/weighted_score['weighted_help']
    weighted_score['sentiment_average_score'] = weighted_score['weighted_sentiment_average']/weighted_score['weighted_help']
    weighted_score['sentiment_rescaled'] = weighted_score['weighted_sentiment_rescaled']/weighted_score['weighted_help']
    weighted_score['overall_score'] = weighted_score['weighted_overall']/weighted_score['weighted_help']
    weighted_score['combined_score'] = weighted_score['weighted_combined']/weighted_score['weighted_help']
    
    return weighted_score[['sentiment_review_score', 'sentiment_summary_score', 'sentiment_average_score', 'sentiment_rescaled', 'overall_score', 'combined_score']]

In [34]:
matched_ebooks= pd.read_csv(matched_ebooks_path)
matched_books = pd.read_csv(matched_books_path)

In [35]:
if WRITE_WEIGHTED_SCORE:
    weighted_scores_books = weighted_scores(matched_books.copy())
    weighted_scores_ebooks = weighted_scores(matched_ebooks.copy())

    weighted_scores_books.to_csv(weighted_scores_books_path)
    weighted_scores_ebooks.to_csv(weighted_scores_ebooks_path)

weighted_scores_books = pd.read_csv(weighted_scores_books_path)
weighted_scores_ebooks = pd.read_csv(weighted_scores_ebooks_path)

Once we have computed the different scores for the books and the ebooks, it is now time to combine those scores with our data.

In [36]:
merged_with_books = pd.merge(merged_metadatas, weighted_scores_books, left_on='asin_book', right_on='asin')
merged_with_all = pd.merge(merged_with_books, weighted_scores_ebooks, left_on='asin_ebook', right_on='asin', suffixes=['_book', '_ebook'])

In [37]:
merged_with_all[['title', 'combined_score_book', 'combined_score_ebook']].head()

title  combined_score_book  \
0                                   the ice princess             3.599635   
1                                 the sweetest taboo             3.931882   
2  dragons from the sea (the strongbow saga, book 2)             4.250450   
3                                               iced             4.246412   
4                                      made in italy             4.486103   

   combined_score_ebook  
0              4.257872  
1              4.492361  
2              4.272470  
3              4.029660  
4              4.065768

In [38]:
merged_with_all.shape[0]

53

In [39]:
average_combined_score = merged_with_all[['combined_score_book', 'combined_score_ebook']].mean()
average_combined_score

combined_score_book     4.049749
combined_score_ebook    4.108649
dtype: float64

In [40]:
average_sentiment_score = merged_with_all[['sentiment_average_score_book', 'sentiment_average_score_ebook']].mean()
average_sentiment_score

sentiment_average_score_book     0.454737
sentiment_average_score_ebook    0.462108
dtype: float64

In [41]:
average_overall_score = merged_with_all[['overall_score_book', 'overall_score_ebook']].mean()
average_overall_score

overall_score_book     4.190025
overall_score_ebook    4.293082
dtype: float64

Here we are going to see if the sentiment's score is consistent with the ratings, meaning that reviewers giving high ratings should also be positive in their reviews.

In [42]:
reviews_books_with_sentiment = matched_books.copy()
f_sentiment = lambda x: sid.polarity_scores(x)['compound']
reviews_books_with_sentiment['sentiment_review'] = matched_books['reviewText'].apply(f_sentiment)
reviews_books_with_sentiment['sentiment_summary'] = matched_books['summary'].apply(f_sentiment)
reviews_books_with_sentiment['average'] = (reviews_books_with_sentiment['sentiment_review'] + reviews_books_with_sentiment['sentiment_summary'])*0.5

In [43]:
reviews_books_with_sentiment.groupby('overall').mean()

Unnamed: 0  sentiment_review  sentiment_summary   average
overall                                                             
1        3.683335e+06          0.212659          -0.160626  0.026016
2        3.191242e+06          0.258701          -0.034349  0.112176
3        3.451677e+06          0.467796           0.138467  0.303131
4        3.435581e+06          0.675070           0.279179  0.477124
5        3.741658e+06          0.728469           0.356776  0.542622

In [44]:
reviews_books_with_sentiment[reviews_books_with_sentiment['overall'] == 1].head()

Unnamed: 0        asin  overall                             summary  \
11        23417  0007269854        1            Absurd police ineptitude   
64        23470  0007269854        1                            Hated it   
65        23471  0007269854        1                         Not so good   
67        23473  0007269854        1            Audiobook CDs poor grade   
100       23506  0007269854        1  I don't usually write bad reviews.   

         reviewerID   helpful  \
11   A38L3I3R3VRPTT    [0, 0]   
64   A2T1ZGMIFSHWW2    [5, 5]   
65   A23J25BQYRAIVX  [13, 15]   
67   A3UYDNMNB7VX75    [0, 0]   
100  A3RIGC6OUSKQ8R    [6, 6]   

                                            reviewText  \
11   I agree with the complaints about cliches, car...   
64   Giving up at pg 139 (of 389, paperback). I no ...   
65   If you're a fan of Scandinavian crime fiction,...   
67   My complaint is not with the novel, but with t...   
100  However, I am 78 pages into this book and I ma...   

                          reviewerName  sentiment_review  sentiment_summary  \
11   Bay Area Bibliophile "marisylvia"            0.6467             0.0000   
64                     In Vino Veritas            0.8126            -0.6369   
65                        Jack Tierney           -0.9862             0.4927   
67          Jane F. Wiedel "dog lover"           -0.9821            -0.4767   
100                     Mary Ann Moore           -0.7228             0.4310   

     average  
11   0.32335  
64   0.08785  
65  -0.24675  
67  -0.72940  
100 -0.14590

In [45]:
reviews_books_with_sentiment.iloc[212]['reviewText']

"This book is just plain horrible and for me to say that is saying a lot.  I read everything and a Danielle Steel or Nora Roberts novel is like Shakespeare compared to this book.  I picked this book up because I loved Michael Chriton's Timeline so much.  I am facinated with the medieval ages and think the concept of someone from today's time comming in contact with the past is an intresting concept.  Unfortunately this book is not intresting in the least.  I went into the book really wanting to like it but the writing is so stiff and awkard.  The dialogue is not believable in the least.  The romance between Edward and Robyn which should be the focal point of the book is boring.  Edward is like a Ken doll, looks cute but is plastic with no personality.  I guess it doesn't matter in the end because he is not around for most of the story anyway.  If you read this book thinking you will learn something about the medieval ages you will be sorely dissappointed.  Edward doesn't talk like he's

This review above has a really positive sentiment score of 0.9735, but if we read the review, we can understand that it is quite the opposite. A second example is "Not so good" which yields a score of 0.4927, showing us the limitation of Vader.

We can observe here that the better the rating is, the more positive the text of the review is in average. The reviews with a rating of 1 are not negative in average. Even if the scores are not always accurate due to the limitation of Vader, they are consistent in average with the ratings.

In [46]:
reviews_ebooks_with_sentiment = matched_ebooks.copy()
reviews_ebooks_with_sentiment['sentiment_review'] = matched_ebooks['reviewText'].apply(f_sentiment)
reviews_ebooks_with_sentiment['sentiment_summary'] = matched_ebooks['summary'].apply(f_sentiment)
reviews_ebooks_with_sentiment['average'] = (reviews_ebooks_with_sentiment['sentiment_review'] + reviews_ebooks_with_sentiment['sentiment_summary'])*0.5

In [47]:
reviews_ebooks_with_sentiment.groupby('overall').mean()

Unnamed: 0  sentiment_review  sentiment_summary   average
overall                                                             
1        23266.285714         -0.150765          -0.153098 -0.151932
2        28643.590909          0.053472          -0.045726  0.003873
3        28790.469697          0.503538           0.186164  0.344851
4        29388.108069          0.663903           0.268574  0.466238
5        27860.713587          0.688401           0.355370  0.521885

The ratings seem to be consistent with the sentiment's score here again. Compared to books, the sentiment's score is slightly lower and even negative for a rating of 1.

For the same rating, the books reviewer seems to be more positive in their review, this may be explained by the fact that books reviews tend to be longer and more descriptive than ebooks reviews.

This analysis may not be really representative of the reality due to the limitation in size of our final dataset.

Sources:
* *Improving the Amazon Review System by Exploiting the Credibility and Time-Decay of Public Reviews*, Bo-Chun Wang, Wen-Yuan Zhu, and Ling-Jyh Chen
* Hutto, C.J. & Gilbert, E.E. (2014). *VADER: A Parsimonious Rule-based Model for Sentiment Analysis of Social Media Text. Eighth International Conference on Weblogs and Social Media (ICWSM-14)*. Ann Arbor, MI, June 2014.